src/repository/SchedulesRepository.ts

```diff
@@ -25,97 +27,127 @@ export class SchedulesRepository extends Repository<ScheduleEntity> {
     this.logger = logger;
   }
 
+  getLogData(): MomoEventData {
+    return { name: this.name, scheduleId: this.scheduleId };
+  }
+
+  async deleteOne(): Promise<void> {
+    await this.collection.deleteOne({ scheduleId: this.scheduleId });
+  }
+
   /**
-   * Checks if there is an alive active schedule in the database for the given name.
+   * Checks the state of the schedule represented by this repository.
+   *
+   * INACTIVE: There is currently no active instance for a schedule with this name.
+   * DIFFERENT_INSTANCE_ACTIVE: Another instance (but not this one) is active for the schedule with this name.
+   * THIS_INSTANCE_ACTIVE: This instance is active for the schedule with this name.
    *
-   * If there is one -> return true if it is this schedule, otherwise false.
-   * If there is no such schedule -> we try inserting this schedule as the active one.
-   * ↳ If it worked return true, otherwise false.
+   * @param now timestamp in milliseconds
+   * @returns the schedule's state
    */
-  async isActiveSchedule(name: string): Promise<boolean> {
-    const lastAlive = DateTime.now().toMillis();
-    const threshold = lastAlive - this.deadScheduleThreshold;
+  async isActiveSchedule(now: number): Promise<boolean> {
+    const threshold = now - this.deadScheduleThreshold;
+    const activeSchedule = await this.collection.findOne({ name: this.name });
+
+    if (activeSchedule === null || activeSchedule.scheduleId === this.scheduleId) {
+      return true;
+    }
+
+    return activeSchedule.lastAlive < threshold; // if activeSchedule is too old, take over and make this one active
+  }
+
+  /**
+   * Tries to set this instance as active
+   *
+   * @param now timestamp in milliseconds
+   * @returns true if this instance is now active for the schedule with this name, false otherwise
+   */
+  async setActiveSchedule(now: number): Promise<boolean> {
+    const threshold = now - this.deadScheduleThreshold;
+
+    const deadSchedule: Filter<ScheduleEntity> = { name: this.name, lastAlive: { $lt: threshold } };
+    const thisSchedule: Filter<ScheduleEntity> = { scheduleId: this.scheduleId };
+
+    const updatedSchedule: ScheduleEntity = {
+      name: this.name,
+      scheduleId: this.scheduleId,
+      lastAlive: now,
+      executions: {},
+    };
+
     try {
-      const result = await this.collection.findOneAndUpdate(
-        { lastAlive: { $lt: threshold }, name },
-        {
-          $set: {
-            name,
-            scheduleId: this.scheduleId,
-            lastAlive,
-            executions: {},
-          },
-        },
-        {
-          upsert: true,
-          returnDocument: 'after',
-        },
+      await this.collection.updateOne(
+        { $or: [deadSchedule, thisSchedule] },
```

Comment:
I don't quite get this $or clause.

Reply:
We want to update in exactly two cases:
there is an active schedule, but it is dead and will be replaced by me
or: I am the active schedule currently and only will update the timestamp

Any hints how to make this clearer are welcome. Maybe I can add some words to the jsdoc.

Reply:
I added a comment and extended the jsdoc of the method. Is it clearer now?

Reply:
Awesome, thanks! Looks good!